Add PUT and POST Endpoints for SSP Implemented Requirement

hreineck opened the pull request:
Add documentation for a new endpoint for a PUT request that updates
an implemented requirement in an SSP.

Closes EasyDynamics/easygrc#19

hreineck committed:
Add PUT Endpoint for SSP Implemented Requirement

Add documentation for a new endpoint for a PUT request that updates
an implemented requirement in an SSP.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/EasyDynamics/oscal-rest?pullRequest=47) <br/>All committers have signed the CLA.

hreineck committed:
Add POST Endpoint for Adding Implemented Requirement to SSP

Add the documentation for a new endpoint for a POST request that adds
an implemented requirement to an SSP.

pkothare commented on openapi.yaml:
```diff
@@ -1844,6 +1844,68 @@ paths:
+        - oscal_auth:
+            - write:ssps
+            - read:ssps
+  /system-security-plans/{sspId}/control-implementation/implemented-requirements/{implementedRequirementID}:
```
```suggestion
  /system-security-plans/{sspId}/control-implementation/implemented-requirements/{implementedRequirementId}:
```
.. Just to keep naming consistent.

pkothare commented on openapi.yaml:
```diff
@@ -1844,6 +1844,68 @@ paths:
+          required: true
+          schema:
+            type: string
+        - name: implementedRequirementID
```
```suggestion
        - name: implementedRequirementId
```

pkothare commented on openapi.yaml:
```diff
@@ -1844,6 +1844,68 @@ paths:
+      summary: Add an implemented requirement to an OSCAL system security plan
+      operationId: addImplementedRequirementToSsp
+      parameters:
+        - name: api_key
```
Are we including `api_key` on all operations as part of the spec? Maybe @rgauss you can give me a quick yes/no on this.

rgauss replied:
This is really a holdover example.  This and things like the `oscal_auth` references need to be revamped at some point.

In short, no, we don't need it.

rgauss replied:
@hreineck, can we remove these?

pkothare reviewed (changes requested):
Requesting minor tweaks based on the naming convention we've used.

rgauss reviewed (changes requested):
We'll need to specify what `requestBody` looks like on these as well.

See the [OSCALParty schema](https://github.com/EasyDynamics/oscal-rest/blob/develop/openapi.yaml#L2097) for precedence on how to pick out a 'subobject' from a root OSCAL object.

hreineck committed:
Update Request Body for New Endpoints

Create a definition of the `OSCALImplementedRequirement` schema to be
referenced by the new PUT and POST endpoints, and add a `requestBody`
to document the usage of these endpoints; the body must contain a
OSCAL Implemented Requirement object as described in the schema.

hreineck committed:
Update Naming of ID Property

Co-authored-by: Pranav A. Kothare <pkothare@easydynamics.com>

rgauss commented on openapi.yaml:
```diff
@@ -1844,6 +1844,82 @@ paths:
+          required: true
+          schema:
+            type: string
+      responses:
```
Should we add a `201` response here?

rgauss commented on openapi.yaml:
```diff
@@ -1844,6 +1844,82 @@ paths:
+          required: true
+          schema:
+            type: string
+      responses:
```
Should we add a `200` here?

rgauss commented on openapi.yaml:
```diff
@@ -2088,6 +2164,10 @@ components:
+    OSCALImplementedRequirement:
+      type: object
+      $ref:
+        "https://raw.githubusercontent.com/usnistgov/OSCAL/main/json/schema/oscal_ssp_schema.json#assembly_oscal-ssp_implemented-requirement"
```
This currently needs to reference the `EasyDynamics` fork like the other schema refs (though NIST may be fixing that soon).

You can double check by trying to preview this file in the online Swagger editor.

hreineck committed:
Remove api-key Parameters

Remove the unnecessary `api-key` parameters from the new endpoints.

hreineck committed:
Add Success Responses

Add the 200/201 responses to the new endpoints.

hreineck committed:
Fix Schema Link

Change the `OSCALImplementedRequirement` schema reference link to
the EasyDynamics fork so that it can be used in the documentation.

hreineck committed:
Fix Indentation Error

Resolve a linting issue with indentation

mikeisen1 reviewed (changes requested):
Everything looks good except we need a valid reference to the implemented-requirements schema.

mikeisen1 commented on openapi.yaml:
```diff
@@ -2088,6 +2170,10 @@ components:
+    OSCALImplementedRequirement:
+      type: object
+      $ref:
+        "https://raw.githubusercontent.com/EasyDynamics/OSCAL/main/json/schema/oscal_ssp_schema.json#assembly_oscal-ssp_implemented-requirement"
```
```suggestion
    OSCALImplementedRequirement:
      type: object
      properties:
        implemented-requirement:
          $ref:
            "https://raw.githubusercontent.com/EasyDynamics/OSCAL/json-schema-ref-by-path/json/schema/oscal_ssp_schema.json#/definitions/assembly_oscal-ssp_implemented-requirement"
```

hreineck committed:
Fix Schema Reference

Correct the reference to the OSCAL Implemented Requirement schema

Co-authored-by: Michael Eisenberg <43248318+mikeisen1@users.noreply.github.com>

rgauss commented on openapi.yaml:
```diff
@@ -1844,6 +1844,88 @@ paths:
+    put:
+      tags:
+        - OSCAL System Security Plan
+      summary: Updates the body of an implemented requirement in an OSCAL system security plan
```
```suggestion
      summary: Updates an implemented requirement in an OSCAL system security plan
```

rgauss reviewed (approved):
Minor suggestion on description that does not have to hold up review.

rgauss committed:
Update openapi.yaml

rgauss commented:
@pkothare, does this look good to you now?
